packages/playwright-test/src/worker/workerMain.ts

```diff
@@ -444,43 +441,44 @@ export class WorkerMain extends ProcessRunner {
-          this._didRunFullCleanup = true;
+      if (this._isStopped) {
+        // Run all remaining "afterAll" hooks and teardown all fixtures when worker is shutting down.
+        // Mark as "cleaned up" early to avoid running cleanup twice.
+        this._didRunFullCleanup = true;
 
-          // Give it more time for the full cleanup.
-          await testInfo._runWithTimeout(async () => {
-            debugTest(`running full cleanup after the failure`);
+        // Give it more time for the full cleanup.
+        await testInfo._runWithTimeout(async () => {
+          debugTest(`running full cleanup after the failure`);
 
-            const teardownSlot = { timeout: this._project.project.timeout, elapsed: 0 };
-            // Attribute to 'test' so that users understand they should probably increase the test timeout to fix this issue.
-            await testInfo._timeoutManager.runRunnable({ type: 'test', slot: teardownSlot }, async () => {
-              debugTest(`tearing down test scope started`);
-              const testScopeError = await testInfo._runAndFailOnError(() => this._fixtureRunner.teardownScope('test', testInfo._timeoutManager));
-              debugTest(`tearing down test scope finished`);
-              firstAfterHooksError = firstAfterHooksError || testScopeError;
+          const teardownSlot = { timeout: this._project.project.timeout, elapsed: 0 };
+          // Attribute to 'test' so that users understand they should probably increase the test timeout to fix this issue.
+          await testInfo._timeoutManager.runRunnable({ type: 'test', slot: teardownSlot }, async () => {
+            debugTest(`tearing down test scope started`);
+            const testScopeError = await testInfo._runAndFailOnError(() => this._fixtureRunner.teardownScope('test', testInfo._timeoutManager));
+            debugTest(`tearing down test scope finished`);
+            firstAfterHooksError = firstAfterHooksError || testScopeError;
+          });
 
-              for (const suite of reversedSuites) {
-                const afterAllError = await this._runAfterAllHooksForSuite(suite, testInfo);
-                firstAfterHooksError = firstAfterHooksError || afterAllError;
-              }
+          for (const suite of reversedSuites) {
+            const afterAllError = await this._runAfterAllHooksForSuite(suite, testInfo);
+            firstAfterHooksError = firstAfterHooksError || afterAllError;
+          }
 
-              debugTest(`tearing down worker scope started`);
-              const workerScopeError = await testInfo._runAndFailOnError(() => this._fixtureRunner.teardownScope('worker', testInfo._timeoutManager));
-              debugTest(`tearing down worker scope finished`);
-              firstAfterHooksError = firstAfterHooksError || workerScopeError;
-            });
+          await testInfo._timeoutManager.runRunnable({ type: 'teardown', slot: teardownSlot }, async () => {
```

Comment:
Please bring back the comment from previous PR that said `Attribute to 'teardown'...`